ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/LedgerApiServer.scala

```diff
@@ -52,136 +65,99 @@ object LedgerApiServer {
+      mat.system.name + "-nio-worker",
+      parallelism = Runtime.getRuntime.availableProcessors)
+    val bossEventLoopGroup = createEventLoopGroup(mat.system.name + "-nio-boss", parallelism = 1)
+    closeables.push(() => {
+      // `shutdownGracefully` has a "quiet period" which specifies a time window in which
+      // _no requests_ must be witnessed before shutdown is _initiated_. Here we want to
+      // start immediately, so no quiet period -- by default it's 2 seconds.
+      // Moreover, there's also a "timeout" parameter
+      // which caps the time to wait for the quiet period to be fullfilled. Since we have
+      // no quiet period, this can also be 0.
+      // See <https://netty.io/4.1/api/io/netty/util/concurrent/EventExecutorGroup.html#shutdownGracefully-long-long-java.util.concurrent.TimeUnit->.
+      // The 10 seconds to wait is sort of arbitrary, it's long enough to be noticeable though.
+      Seq(
+        workerEventLoopGroup.shutdownGracefully(0, 0, MILLISECONDS),
+        bossEventLoopGroup.shutdownGracefully(0, 0, MILLISECONDS),
+      ).foreach(_.await(10, SECONDS))
+    })
 
     createApiServices(mat, serverEsf).map { apiServices =>
-      new ApiServer {
-        private val impl = new LedgerApiServer(
-          apiServices,
-          desiredPort,
-          maxInboundMessageSize,
-          address,
-          loggerFactory,
-          sslContext,
-          interceptors,
-          metrics
-        )
-
-        /** returns the api port the server is listening on */
-        override def port: Int = impl.port
-
-        /** returns when all services have been closed during the shutdown */
-        override def servicesClosed(): Future[Unit] = impl.servicesClosed()
-
-        override def close(): Unit = {
-          impl.close()
-          serverEsf.close()
+      val builder = address.fold(NettyServerBuilder.forPort(desiredPort))(address =>
+        NettyServerBuilder.forAddress(new InetSocketAddress(address, desiredPort)))
+      sslContext
+        .fold {
+          logger.info("Starting plainText server")
+        } { sslContext =>
+          logger.info("Starting TLS server")
+          val _ = builder.sslContext(sslContext)
         }
+      builder.directExecutor()
+      builder.channelType(classOf[NioServerSocketChannel])
+      builder.bossEventLoopGroup(bossEventLoopGroup)
+      builder.workerEventLoopGroup(workerEventLoopGroup)
+      builder.permitKeepAliveTime(10, SECONDS)
+      builder.permitKeepAliveWithoutCalls(true)
+      builder.maxInboundMessageSize(maxInboundMessageSize)
+      interceptors.foreach(builder.intercept)
+      builder.intercept(new MetricsInterceptor(metrics))
+      apiServices.services.foreach(builder.addService)
+      val grpcServer = builder.build()
+
+      try {
+        grpcServer.start()
+      } catch {
+        case io: IOException if io.getCause != null && io.getCause.isInstanceOf[BindException] =>
+          stop()
```

Comment:
If you put `stop()` in the `finally` block it's guaranteed to be executed even if you re-throw. Any reason to not put it there and skip the `NonFatal` rethrow?

Reply:
Done.